1-Draft/RFCNNNN-String-Manipulating-Cmdlets.md

```diff
@@ -0,0 +1,84 @@
+---
+RFC: RFCNNNN
+Author: Friedrich Weinmann
+Status: Draft
+SupercededBy: n/a
+Version: 1.0
+Area: cmdlets/data manipulation
+Comments Due: 
+Plan to implement: Yes
+---
+
+# String Manipulating Cmdlets
+
+Add cmdlets that manipulate strings and are capable of doing so on the pipeline.
+
+## Motivation
+
+Existing tools to manipulate strings - the most common data type manipulated in PowerShell are lacking in usability.
+The .NET tools available through string and regex are inconvenient to use.
+String-manipulating operators (such as `-split`) are unintuitive to explore, as their options are not easily discoverable.
+Neither option allows use on the pipeline, forcing interruption of the pipeline or use of inefficient `ForEach-Object` calls.
+
+From this come two main pain points this RFC is aimed to resolve:
+
+Current string data manipulation ...
+
+ - is technically insufficient from a performance perspective.
+ - is unintuitive and inconvenient for the end user.
+
+Given that this is a universal problem affecting literally all users and most code, this should be implemented as part of the core application, not an external third party module.
+
+## Specification
+
+### New Cmdlets
+
+The solution should provide cmdlets that allow string manipulation on the pipeline.
+
+The following operations should be supported:
+
+ - `Add-String`    | Implements partial `-f` functionality to add content to strings, as well as the `.PadLeft()` and `.PadRight()` methods.
+ - `Format-String` | Implements full `-f` functionality or the `String.Format()` method. Supports gathering multiple items from the pipeline before formatting.
+ - `Get-SubString` | Implements the `.SubString()` method, the `-Trim` operator as well as the `.TrimStart()` and `.TrimEnd()` methods.
+ - `Join-String`   | Implements the `-join` operator, allowing specifying the number of items to join in each batch.
+ - `Set-String`    | Implements the `-Replace` operator as well as the `.Replace()` method.
+ - `Split-String`  | Implements the `-Split` operator as well as the `.Split()` method.
+
+### New Aliases
+
+Given that part of the goal is to improve the user experience, introducing aliases - especially for those mapping operators where users are used to it - will improve user experience and accelerate adoption.
+
+Thus here a proposed list of aliases:
+
+ - `Add-String`    | `wrap` or `add`
+ - `Format-String` | `format`
+ - `Get-SubString` | None needed - `substring` will automatically resolve to `Get-SubString`
+ - `Join-String`   | `join`
+ - `Set-String`    | `replace`
+ - `Split-String`  | `split`
+
```

Comment:
Re `substring` automatically resolving to `Get-SubString`: This comes with a hefty performance penalty (and the implementation is incomplete), [leading @lzybkr to ponder removing the feature altogether](https://github.com/PowerShell/PowerShell/issues/3987#issuecomment-308279494).

Additionally, we have to be careful not to clash with preinstalled external utilities, which on Unix-like platforms applies to `join` and `split`. 

Unfortunately, using the sanctioned verb alias prefixes and abbreviating noun `String` as `s` does not make for descriptive names (and can also clash): `Add-String` -> `as` (clashes with `/usr/bin/as` (assembler) on Unix); `Format-String` -> `fs`, `Get-Substring` -> `gss`, `Join-String` -> `js`, `Set-String` -> `ss` (clashes with `/bin/ss` (socket utility) on Linux), `Split-String` -> `sls` (yes, that would clash with `Select-String`'s alias, which is irregularly named and should be `scs`)

Not sure what the right answer is; stick `str` in the names? `addstr`, `formatstr`, `substr`, `joinstr`, `setstr`, `splitstr`?

Reply:
Well, `sls => Select-String` clearly establishes precedence for `s`  as string, but you might use `str` to disambiguate when necessary...